## Data Transformation
***Clean data, remove duplicates, sort anomalies and make it healthy for analysis***

In [202]:
# Import neccessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [203]:
# Load data from local storage into a pandas dataframe
htb = pd.read_csv(r"C:/Users/HP/Documents/Azeez/Hotel Booking Analysis/Hotel Bookings.csv")

In [204]:
# Set all columns to show
pd.set_option('display.max_columns', None)

# Preview data
htb.head()

s/n         hotel  is_canceled  lead_time  arrival_date_year  \
0    1  Resort Hotel            0        342               2015   
1    2  Resort Hotel            0        737               2015   
2    3  Resort Hotel            0          7               2015   
3    4  Resort Hotel            0         13               2015   
4    5  Resort Hotel            0         14               2015   

  arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  \
0                        0                     0       2       0.0       0   
1                        0                     0       2       0.0       0   
2                        0                     1       1       0.0       0   
3                        0                     1       1       0.0       0   
4                        0                     2       2       0.0       0   

  meal country market_segment distribution_channel  is_repeated_guest  \
0   BB     PRT         Direct               Direct                  0   
1   BB     PRT         Direct               Direct                  0   
2   BB     GBR         Direct               Direct                  0   
3   BB     GBR      Corporate            Corporate                  0   
4   BB     GBR      Online TA                TA/TO                  0   

   previous_cancellations  previous_bookings_not_canceled reserved_room_type  \
0                       0                               0                  C   
1                       0                               0                  C   
2                       0                               0                  A   
3                       0                               0                  A   
4                       0                               0                  A   

  assigned_room_type  booking_changes deposit_type  agent  company  \
0                  C                3   No Deposit    NaN      NaN   
1                  C                4   No Deposit    NaN      NaN   
2                  C                0   No Deposit    NaN      NaN   
3                  A                0   No Deposit  304.0      NaN   
4                  A                0   No Deposit  240.0      NaN   

   days_in_waiting_list customer_type   adr  required_car_parking_spaces  \
0                     0     Transient   0.0                            0   
1                     0     Transient   0.0                            0   
2                     0     Transient  75.0                            0   
3                     0     Transient  75.0                            0   
4                     0     Transient  98.0                            0   

   total_of_special_requests reservation_status reservation_status_date  
0                          0          Check-Out                7/1/2015  
1                          0          Check-Out                7/1/2015  
2                          0          Check-Out                7/2/2015  
3                          0          Check-Out                7/2/2015  
4                          1          Check-Out                7/3/2015

In [205]:
# Back up data into another dataframe
htb_backup = htb.copy()

In [206]:
# Inspect the data
htb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 33 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   s/n                             119390 non-null  int64  
 1   hotel                           119390 non-null  object 
 2   is_canceled                     119390 non-null  int64  
 3   lead_time                       119390 non-null  int64  
 4   arrival_date_year               119390 non-null  int64  
 5   arrival_date_month              119390 non-null  object 
 6   arrival_date_week_number        119390 non-null  int64  
 7   arrival_date_day_of_month       119390 non-null  int64  
 8   stays_in_weekend_nights         119390 non-null  int64  
 9   stays_in_week_nights            119390 non-null  int64  
 10  adults                          119390 non-null  int64  
 11  children                        119386 non-null  float64
 12  babies          

In [207]:
# Check for null values
htb.isna().sum()

s/n                                    0
hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent           

In [208]:
# Check no of rows and columns
rows, cols = htb.shape

print(f'The \'htb\' dataframe has {rows} rows and {cols} columns.')

The 'htb' dataframe has 119390 rows and 33 columns.


In [209]:
htb

s/n         hotel  is_canceled  lead_time  arrival_date_year  \
0            1  Resort Hotel            0        342               2015   
1            2  Resort Hotel            0        737               2015   
2            3  Resort Hotel            0          7               2015   
3            4  Resort Hotel            0         13               2015   
4            5  Resort Hotel            0         14               2015   
...        ...           ...          ...        ...                ...   
119385  119386    City Hotel            0         23               2017   
119386  119387    City Hotel            0        102               2017   
119387  119388    City Hotel            0         34               2017   
119388  119389    City Hotel            0        109               2017   
119389  119390    City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
0                    July                        27   
1                    July                        27   
2                    July                        27   
3                    July                        27   
4                    July                        27   
...                   ...                       ...   
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  children  babies meal country  \
0                          0       2       0.0       0   BB     PRT   
1                          0       2       0.0       0   BB     PRT   
2                          1       1       0.0       0   BB     GBR   
3                          1       1       0.0       0   BB     GBR   
4                          2       2       0.0       0   BB     GBR   
...                      ...     ...       ...     ...  ...     ...   
119385                     5       2       0.0       0   BB     BEL   
119386                     5       3       0.0       0   BB     FRA   
119387                     5       2       0.0       0   BB     DEU   
119388                     5       2       0.0       0   BB     GBR   
119389                     7       2       0.0       0   HB     DEU   

       market_segment distribution_channel  is_repeated_guest  \
0              Direct               Direct                  0   
1              Direct               Direct                  0   
2              Direct               Direct                  0   
3           Corporate            Corporate                  0   
4           Online TA                TA/TO                  0   
...               ...                  ...                ...   
119385  Offline TA/TO                TA/TO                  0   
119386      Online TA                TA/TO                  0   
119387      Online TA                TA/TO                  0   
119388      Online TA                TA/TO                  0   
119389      Online TA                TA/TO                  0   

        previous_cancellations  previous_bookings_not_canceled  \
0   

**Data Facts**

The dataset contains 119,390 rows and 33 columns.

Missing values were found in the following columns:

- Children: 4 null values
- Country: 488 null values
- Agent: 16,340 null values
- Company: 112,593 null values

Data ranges from 2013 to 2017 (covers 5 years of booking records).

### **Data Cleaning**
1. Get 'Booking Date'
    
    NB: Lead time the number of days between Booking and Arrival (i.e. 'Arrival Date' minus 'Booking Date')

    To get 'Booking Date', we need to bring our 'Arrival Date' together

In [210]:
# Join the Arrival Date Columns, and convert to DateTime
htb['arrival_date'] = pd.to_datetime(
    htb['arrival_date_year'].astype(str) + '-' + 
    htb['arrival_date_month'].astype(str) + '-' + 
    htb['arrival_date_day_of_month'].astype(str), 
    format="%Y-%B-%d"
    )

# Drop the 'arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month', 'arrival_date_week_number'
htb = htb.drop(columns=['arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month', 'arrival_date_week_number'])

# Preview data
htb.head()

s/n         hotel  is_canceled  lead_time  stays_in_weekend_nights  \
0    1  Resort Hotel            0        342                        0   
1    2  Resort Hotel            0        737                        0   
2    3  Resort Hotel            0          7                        0   
3    4  Resort Hotel            0         13                        0   
4    5  Resort Hotel            0         14                        0   

   stays_in_week_nights  adults  children  babies meal country market_segment  \
0                     0       2       0.0       0   BB     PRT         Direct   
1                     0       2       0.0       0   BB     PRT         Direct   
2                     1       1       0.0       0   BB     GBR         Direct   
3                     1       1       0.0       0   BB     GBR      Corporate   
4                     2       2       0.0       0   BB     GBR      Online TA   

  distribution_channel  is_repeated_guest  previous_cancellations  \
0               Direct                  0                       0   
1               Direct                  0                       0   
2               Direct                  0                       0   
3            Corporate                  0                       0   
4                TA/TO                  0                       0   

   previous_bookings_not_canceled reserved_room_type assigned_room_type  \
0                               0                  C                  C   
1                               0                  C                  C   
2                               0                  A                  C   
3                               0                  A                  A   
4                               0                  A                  A   

   booking_changes deposit_type  agent  company  days_in_waiting_list  \
0                3   No Deposit    NaN      NaN                     0   
1                4   No Deposit    NaN      NaN                     0   
2                0   No Deposit    NaN      NaN                     0   
3                0   No Deposit  304.0      NaN                     0   
4                0   No Deposit  240.0      NaN                     0   

  customer_type   adr  required_car_parking_spaces  total_of_special_requests  \
0     Transient   0.0                            0                          0   
1     Transient   0.0                            0                          0   
2     Transient  75.0                            0                          0   
3     Transient  75.0                            0                          0   
4     Transient  98.0                            0                          1   

  reservation_status reservation_status_date arrival_date  
0          Check-Out                7/1/2015   2015-07-01  
1          Check-Out                7/1/2015   2015-07-01  
2          Check-Out                7/2/2015   2015-07-01  
3          Check-Out                7/2/2015   2015-07-01  
4          Check-Out                7/3/2015   2015-07-01

In [211]:
# New column: Subtract 'Lead Time' from 'Arrival Date' 
htb['booking_date'] = htb['arrival_date'] - pd.to_timedelta(htb['lead_time'], unit='D')

# Preview data
htb.head()

s/n         hotel  is_canceled  lead_time  stays_in_weekend_nights  \
0    1  Resort Hotel            0        342                        0   
1    2  Resort Hotel            0        737                        0   
2    3  Resort Hotel            0          7                        0   
3    4  Resort Hotel            0         13                        0   
4    5  Resort Hotel            0         14                        0   

   stays_in_week_nights  adults  children  babies meal country market_segment  \
0                     0       2       0.0       0   BB     PRT         Direct   
1                     0       2       0.0       0   BB     PRT         Direct   
2                     1       1       0.0       0   BB     GBR         Direct   
3                     1       1       0.0       0   BB     GBR      Corporate   
4                     2       2       0.0       0   BB     GBR      Online TA   

  distribution_channel  is_repeated_guest  previous_cancellations  \
0               Direct                  0                       0   
1               Direct                  0                       0   
2               Direct                  0                       0   
3            Corporate                  0                       0   
4                TA/TO                  0                       0   

   previous_bookings_not_canceled reserved_room_type assigned_room_type  \
0                               0                  C                  C   
1                               0                  C                  C   
2                               0                  A                  C   
3                               0                  A                  A   
4                               0                  A                  A   

   booking_changes deposit_type  agent  company  days_in_waiting_list  \
0                3   No Deposit    NaN      NaN                     0   
1                4   No Deposit    NaN      NaN                     0   
2                0   No Deposit    NaN      NaN                     0   
3                0   No Deposit  304.0      NaN                     0   
4                0   No Deposit  240.0      NaN                     0   

  customer_type   adr  required_car_parking_spaces  total_of_special_requests  \
0     Transient   0.0                            0                          0   
1     Transient   0.0                            0                          0   
2     Transient  75.0                            0                          0   
3     Transient  75.0                            0                          0   
4     Transient  98.0                            0                          1   

  reservation_status reservation_status_date arrival_date booking_date  
0          Check-Out                7/1/2015   2015-07-01   2014-07-24  
1          Check-Out                7/1/2015   2015-07-01   2013-06-24  
2          Check-Out                7/2/2015   2015-07-01   2015-06-24  
3          Check-Out                7/2/2015   2015-07-01   2015-06-18  
4          Check-Out                7/3/2015   2015-07-01   2015-06-17

2. Replace Null values

From previous findings, 4 columns have null values
- For 'Agent': (The id of the agent who brought or refered user) Replace NULL values with 'No agent'
- For 'Company': Replace NULL values with 'Not provided'
- For 'Country': Replace NULL values with 'Not provided'
- For 'Children': Replace NULL values with zero (0)

In [212]:
# Replace NULL values accordingly
htb['agent'] = htb['agent'].replace(np.nan, 'No agent')
htb[['company', 'country']] = htb[['company', 'country']].replace(np.nan, 'Not provided')
htb['children'] = htb['children'].replace(np.nan, 0).convert_dtypes()
htb

s/n         hotel  is_canceled  lead_time  stays_in_weekend_nights  \
0            1  Resort Hotel            0        342                        0   
1            2  Resort Hotel            0        737                        0   
2            3  Resort Hotel            0          7                        0   
3            4  Resort Hotel            0         13                        0   
4            5  Resort Hotel            0         14                        0   
...        ...           ...          ...        ...                      ...   
119385  119386    City Hotel            0         23                        2   
119386  119387    City Hotel            0        102                        2   
119387  119388    City Hotel            0         34                        2   
119388  119389    City Hotel            0        109                        2   
119389  119390    City Hotel            0        205                        2   

        stays_in_week_nights  adults  children  babies meal country  \
0                          0       2         0       0   BB     PRT   
1                          0       2         0       0   BB     PRT   
2                          1       1         0       0   BB     GBR   
3                          1       1         0       0   BB     GBR   
4                          2       2         0       0   BB     GBR   
...                      ...     ...       ...     ...  ...     ...   
119385                     5       2         0       0   BB     BEL   
119386                     5       3         0       0   BB     FRA   
119387                     5       2         0       0   BB     DEU   
119388                     5       2         0       0   BB     GBR   
119389                     7       2         0       0   HB     DEU   

       market_segment distribution_channel  is_repeated_guest  \
0              Direct               Direct                  0   
1              Direct               Direct                  0   
2              Direct               Direct                  0   
3           Corporate            Corporate                  0   
4           Online TA                TA/TO                  0   
...               ...                  ...                ...   
119385  Offline TA/TO                TA/TO                  0   
119386      Online TA                TA/TO                  0   
119387      Online TA                TA/TO                  0   
119388      Online TA                TA/TO                  0   
119389      Online TA                TA/TO                  0   

        previous_cancellations  previous_bookings_not_canceled  \
0                            0                               0   
1                            0                               0   
2                            0                               0   
3                            0                               0   
4                            0                               0   
...                        ...                             ...   
119385                       0                               0   
119386                       0                               0   
119387                       0                               0   
119388                       0                               0   
119389                       0                               0   

       reserved_room_type assigned_room_type  booking_changes deposit_type  \
0                       C                  C                3   No Deposit   
1                       C                  C                4   No Deposit   
2                       A                  C                0   No Deposit   
3                       A                  A                0   No Deposit   
4                       A                  A                0   No Deposit   
...                   ...                ...              ...          ...   
119385                  A                  A                0   N

In [213]:
# Confirm that there are no NULL values
htb.isna().sum()

s/n                               0
hotel                             0
is_canceled                       0
lead_time                         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
country                           0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
reserved_room_type                0
assigned_room_type                0
booking_changes                   0
deposit_type                      0
agent                             0
company                           0
days_in_waiting_list              0
customer_type                     0
adr                               0
required_car_parking_spaces       0
total_of_special_requests         0
reservation_status          

3. Change datatype for columns

In [214]:
# Check columns datatypes
htb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 31 columns):
 #   Column                          Non-Null Count   Dtype         
---  ------                          --------------   -----         
 0   s/n                             119390 non-null  int64         
 1   hotel                           119390 non-null  object        
 2   is_canceled                     119390 non-null  int64         
 3   lead_time                       119390 non-null  int64         
 4   stays_in_weekend_nights         119390 non-null  int64         
 5   stays_in_week_nights            119390 non-null  int64         
 6   adults                          119390 non-null  int64         
 7   children                        119390 non-null  Int64         
 8   babies                          119390 non-null  int64         
 9   meal                            119390 non-null  object        
 10  country                         119390 non-null  object 

In [215]:
# Change 'reservation_status_date' to datetime
htb['reservation_status_date'] = pd.to_datetime(htb['reservation_status_date'])

# Recheck columns datatype
htb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 31 columns):
 #   Column                          Non-Null Count   Dtype         
---  ------                          --------------   -----         
 0   s/n                             119390 non-null  int64         
 1   hotel                           119390 non-null  object        
 2   is_canceled                     119390 non-null  int64         
 3   lead_time                       119390 non-null  int64         
 4   stays_in_weekend_nights         119390 non-null  int64         
 5   stays_in_week_nights            119390 non-null  int64         
 6   adults                          119390 non-null  int64         
 7   children                        119390 non-null  Int64         
 8   babies                          119390 non-null  int64         
 9   meal                            119390 non-null  object        
 10  country                         119390 non-null  object 

4. Check and Drop Duplicates

In [216]:
# First Check for duplicates
htb[htb.duplicated()]

Empty DataFrame
Columns: [s/n, hotel, is_canceled, lead_time, stays_in_weekend_nights, stays_in_week_nights, adults, children, babies, meal, country, market_segment, distribution_channel, is_repeated_guest, previous_cancellations, previous_bookings_not_canceled, reserved_room_type, assigned_room_type, booking_changes, deposit_type, agent, company, days_in_waiting_list, customer_type, adr, required_car_parking_spaces, total_of_special_requests, reservation_status, reservation_status_date, arrival_date, booking_date]
Index: []

There are no duplicates, hence - skipping this process

5. Drop unneccessary column(s)

In [217]:
# Delete columns that are not needed or useful for analysis
htb = htb.drop(columns=['s/n'])

'''
When deleting columns, it is important to note that columns that are be useful for identifying records
are to be kept even if they are not needed for analysis. 
This helps maintain the data quality and integrity.

Deleting them might lead to data inaccuracy when some operations like 'Deleting duplicates' are done.
'''

"\nWhen deleting columns, it is important to note that columns that are be useful for identifying records\nare to be kept even if they are not needed for analysis. \nThis helps maintain the data quality and integrity.\n\nDeleting them might lead to data inaccuracy when some operations like 'Deleting duplicates' are done.\n"

6. Rename columns for better readabilty

In [218]:
# Rename column adr to average daily rate for better readability
htb = htb.rename(columns={
    'adr': 'avg_daily_rate'
})

7. Replace abbreviations in some columns with their full meanings for better understanding

In [219]:
# Check for unique values for columns where values are to be replaced
print(f'Values in meal column: {htb['meal'].unique()}')
print(f'Values in distribution_channel column: {htb['distribution_channel'].unique()}')

Values in meal column: ['BB' 'FB' 'HB' 'SC' 'Undefined']
Values in distribution_channel column: ['Direct' 'Corporate' 'TA/TO' 'Undefined' 'GDS']


In [220]:
# Replace abbrevation in 'meal', 'distribution_channel' columns respectively
htb['meal'] = htb['meal'].replace({
    'BB': 'Bed & Breakfast',
    'FB': 'Full Board',
    'HB': 'Half Board',
    'SC': 'Self Catering'
})

htb['distribution_channel'] = htb['distribution_channel'].replace({
    'TA/TO': 'Travel Agent/Tour Operator',
    'GDS': 'Global Distribution System'
})

8. After data cleaning, cross-check for duplicates again

In [221]:
# Recheck for duplicates
htb[htb.duplicated()]

hotel  is_canceled  lead_time  stays_in_weekend_nights  \
5       Resort Hotel            0         14                        0   
22      Resort Hotel            0         72                        2   
43      Resort Hotel            0         70                        2   
138     Resort Hotel            1          5                        1   
200     Resort Hotel            0          0                        0   
...              ...          ...        ...                      ...   
119349    City Hotel            0        186                        0   
119352    City Hotel            0         63                        0   
119353    City Hotel            0         63                        0   
119354    City Hotel            0         63                        0   
119373    City Hotel            0        175                        1   

        stays_in_week_nights  adults  children  babies             meal  \
5                          2       2         0       0  Bed & Breakfast   
22                         4       2         0       0  Bed & Breakfast   
43                         3       2         0       0       Half Board   
138                        0       2         0       0  Bed & Breakfast   
200                        1       1         0       0  Bed & Breakfast   
...                      ...     ...       ...     ...              ...   
119349                     3       2         0       0  Bed & Breakfast   
119352                     3       3         0       0  Bed & Breakfast   
119353                     3       3         0       0  Bed & Breakfast   
119354                     3       3         0       0  Bed & Breakfast   
119373                     3       1         0       0  Bed & Breakfast   

       country market_segment        distribution_channel  is_repeated_guest  \
5          GBR      Online TA  Travel Agent/Tour Operator                  0   
22         PRT         Direct                      Direct                  0   
43         ROU         Direct                      Direct                  0   
138        PRT      Online TA  Travel Agent/Tour Operator                  0   
200        GBR      Online TA  Travel Agent/Tour Operator                  0   
...        ...            ...                         ...                ...   
119349     DEU      Online TA  Travel Agent/Tour Operator                  0   
119352     SWE      Online TA  Travel Agent/Tour Operator                  0   
119353     SWE      Online TA  Travel Agent/Tour Operator                  0   
119354     SWE      Online TA  Travel Agent/Tour Operator                  0   
119373     NLD  Offline TA/TO  Travel Agent/Tour Operator                  0   

        previous_cancellations  previous_bookings_not_canceled  \
5                            0                               0   
22                           0                               0   
43                           0                               0   
138                          0                               0   
200                          0                               0   
...                        ...                             ...   
119349                       0                               0   
119352                       0                               0   
119353                       0                               0   
119354                       0                               0   
119373                       0                               0   

       reserved_room_type assigned_room_type  booking_changes deposit_type  \
5                       A                  A                0   No Deposit   
22                      A                  A                1   No Deposit   
43                      E                  E                0   No Deposit   
138                     D                  D                0   No Deposit   
200                     A                  A                0   No Deposit   
...            

In [222]:
# Print no of duplicates before deleting them
print(f'Number of duplicates: {htb.duplicated().sum():,}')

# Remove duplicates from dataframe 'htb'
htb = htb.drop_duplicates()

# Print no of duplicates after deleting them
print(f'Number of duplicates after removal: {htb.duplicated().sum()}')

Number of duplicates: 31,994
Number of duplicates after removal: 0


In [223]:
# Preview Data
htb

hotel  is_canceled  lead_time  stays_in_weekend_nights  \
0       Resort Hotel            0        342                        0   
1       Resort Hotel            0        737                        0   
2       Resort Hotel            0          7                        0   
3       Resort Hotel            0         13                        0   
4       Resort Hotel            0         14                        0   
...              ...          ...        ...                      ...   
119385    City Hotel            0         23                        2   
119386    City Hotel            0        102                        2   
119387    City Hotel            0         34                        2   
119388    City Hotel            0        109                        2   
119389    City Hotel            0        205                        2   

        stays_in_week_nights  adults  children  babies             meal  \
0                          0       2         0       0  Bed & Breakfast   
1                          0       2         0       0  Bed & Breakfast   
2                          1       1         0       0  Bed & Breakfast   
3                          1       1         0       0  Bed & Breakfast   
4                          2       2         0       0  Bed & Breakfast   
...                      ...     ...       ...     ...              ...   
119385                     5       2         0       0  Bed & Breakfast   
119386                     5       3         0       0  Bed & Breakfast   
119387                     5       2         0       0  Bed & Breakfast   
119388                     5       2         0       0  Bed & Breakfast   
119389                     7       2         0       0       Half Board   

       country market_segment        distribution_channel  is_repeated_guest  \
0          PRT         Direct                      Direct                  0   
1          PRT         Direct                      Direct                  0   
2          GBR         Direct                      Direct                  0   
3          GBR      Corporate                   Corporate                  0   
4          GBR      Online TA  Travel Agent/Tour Operator                  0   
...        ...            ...                         ...                ...   
119385     BEL  Offline TA/TO  Travel Agent/Tour Operator                  0   
119386     FRA      Online TA  Travel Agent/Tour Operator                  0   
119387     DEU      Online TA  Travel Agent/Tour Operator                  0   
119388     GBR      Online TA  Travel Agent/Tour Operator                  0   
119389     DEU      Online TA  Travel Agent/Tour Operator                  0   

        previous_cancellations  previous_bookings_not_canceled  \
0                            0                               0   
1                            0                               0   
2                            0                               0   
3                            0                               0   
4                            0                               0   
...                        ...                             ...   
119385                       0                               0   
119386                       0                               0   
119387                       0                               0   
119388                       0                               0   
119389                       0                               0   

       reserved_room_type assigned_room_type  booking_changes deposit_type  \
0                       C                  C                3   No Deposit   
1                       C                  C                4   No Deposit   
2                       A                  C                0   No Deposit   
3                       A                  A                0   No Deposit   
4                       A                  A                0   No Deposit   
...            

9. Upload clean data to file

*To avoid repition of data cleaning for subsequent analysis*

In [224]:
# Upload clean data to current directory without index
htb.to_csv('hotel_bookings_clean.csv', index=False)